src/compiler/core.ts

```diff
@@ -2552,4 +2561,168 @@ namespace ts {
+        fileExists(path: string): boolean;
+        directoryExists(path: string): boolean;
+        createDirectory(path: string): void;
+        getCurrentDirectory(): string;
+        getDirectories(path: string): string[];
+        readDirectory(path: string, extensions?: ReadonlyArray<string>, exclude?: ReadonlyArray<string>, include?: ReadonlyArray<string>, depth?: number): string[];
+    }
+
+    export interface CachedHost {
+        writeFile(path: string, data: string, writeByteOrderMark?: boolean): void;
+        fileExists(path: string): boolean;
+        directoryExists(path: string): boolean;
+        createDirectory(path: string): void;
+        getCurrentDirectory(): string;
+        getDirectories(path: string): string[];
+        readDirectory(path: string, extensions?: ReadonlyArray<string>, exclude?: ReadonlyArray<string>, include?: ReadonlyArray<string>, depth?: number): string[];
+        addOrDeleteFileOrFolder(fileOrFolder: string): void;
+        clearCache(): void;
+    }
+
+    export function createCachedHost(host: HostForCaching): CachedHost {
+        const cachedReadDirectoryResult = createMap<FileSystemEntries>();
+        const getCurrentDirectory = memoize(() => host.getCurrentDirectory());
+        const getCanonicalFileName = createGetCanonicalFileName(host.useCaseSensitiveFileNames);
+        return {
+            writeFile,
+            fileExists,
+            directoryExists,
+            createDirectory,
+            getCurrentDirectory,
+            getDirectories,
+            readDirectory,
+            addOrDeleteFileOrFolder,
+            clearCache
+        };
+
+        function toPath(fileName: string) {
+            return ts.toPath(fileName, getCurrentDirectory(), getCanonicalFileName);
+        }
+
+        function getFileSystemEntries(rootDir: string) {
+            const path = toPath(rootDir);
+            const cachedResult = cachedReadDirectoryResult.get(path);
+            if (cachedResult) {
+                return cachedResult;
+            }
+
+            const resultFromHost: FileSystemEntries = {
+                files: host.readDirectory(rootDir, /*extensions*/ undefined, /*exclude*/ undefined, /*include*/["*.*"]) || [],
+                directories: host.getDirectories(rootDir) || []
+            };
+
+            cachedReadDirectoryResult.set(path, resultFromHost);
+            return resultFromHost;
+        }
+
+        function canWorkWithCacheForDir(rootDir: string) {
+            // Some of the hosts might not be able to handle read directory or getDirectories
+            const path = toPath(rootDir);
+            if (cachedReadDirectoryResult.get(path)) {
+                return true;
+            }
+            try {
+                return getFileSystemEntries(rootDir);
+            }
+            catch (_e) {
+                return false;
+            }
+        }
+
+        function fileNameEqual(name1: string, name2: string) {

```

Comment:
I dont agree with that. This allows us to retain the casing (instead of always providing using lower case) of file names even if the host is case-insensitive